doc/man/man1/virgilencrypt.1
.TH "VIRGIL" "1" "August 01, 2015" "Virgil Security CLI (0.3.0)" "BSD General Commands Manual"
.SH NAME
.PP
\f[B]encrypt\f[] \-\- encrypt data for given recipients.
.SH SYNOPSIS
.PP
\f[B]virgil encrypt\f[] [\-\-in \f[I]file\f[]] [\-\-out \f[I]file\f[]]
[\-\-content\-info \f[I]file\f[]] [\-\-recipients \f[I]file\f[]]
\f[I]recipient\f[] \f[I]\&...\f[]
.SH DESCRIPTION
.PP
Encrypt data for given recipients.
Recipient can be represented either by the password, or by the Virgil
Public Key.
.SH OPTIONS
.TP
.B \-i \f[I]file\f[], \-\-in \f[I]file\f[]
Data to be encrypted.
If omitted stdin is used.
.RS
.RE
.TP
.B \-o \f[I]file\f[], \-\-out \f[I]file\f[]
Encrypted data.
If omitted stdout is used.
.RS
.RE
.TP
.B \-c \f[I]file\f[], \-\-content\-info \f[I]file\f[]
Content info \- meta information about encrypted data.
If omitted becomes a part of the encrypted data.
.RS
.RE
.TP
.B \-r \f[I]file\f[], \-\-recipients \f[I]file\f[]
(accepted multiple times) File that contains information about
recipients.
Each line can be either empty line, or comment line, or recipient
defined in format:
.RS
.IP
.nf
\f[C]
[pass|file|email|phone|domain]:<value>
\f[]
.fi
.PP
where:
.IP \[bu] 2
if \f[C]pass\f[], then \f[I]value\f[] \- recipient\[aq]s password;
.IP \[bu] 2
if \f[C]file\f[], then \f[I]value\f[] \- recipient\[aq]s Virgil Public
Key file stored locally;
.IP \[bu] 2
if \f[C]email\f[], then \f[I]value\f[] \- recipient\[aq]s email;
.IP \[bu] 2
if \f[C]phone\f[], then \f[I]value\f[] \- recipient\[aq]s phone;
.IP \[bu] 2
if \f[C]domain\f[], then \f[I]value\f[] \- recipient\[aq]s domain.
.RE
.TP
.B \f[I]recipient\f[]
(accepted multiple times) Contains information about one recipient.
Same as significant line in the recipients configuration file.
.RS
.RE
.SH SEE ALSO
.PP
\f[C]virgil(1)\f[], \f[C]virgildecrypt(1)\f[]
.SH AUTHORS
(c) Virgil Security Inc.


ChangeLog
virgil CLI ChangeLog (Sorted per date)

= Version 0.3.3 released 2015-08-18

Bugfix:
  * Embed MSVC redistributable package installation to the CLI package.

Changes:
  * Add executable folder to the PATH environment variable silently on Windows.

= Version 0.3.2 released 2015-08-17

Bugfix:
  * Fix RSA and EC keys generation.

= Version 0.3.1 released 2015-08-17

Bugfix:
  * Add missing win32 runtime libraries.

= Version 0.3.0 released 2015-08-13

Improvements:
  * Add Markdown documentation.

Changes:
  * Replace certificate term with Virgil Public Key term.
  * Remove options to specify sign's output format.
  * Use C++11 standard.

Features:
  * Add command "keyreg" - register user's public key on the Virgil Public Keys service.
  * Add command "keyget" - get user's public key from the Virgil Public Keys service.
  * Add command "confirm" - send confirmation code to the Virgil Public Keys service.
  * Add command "reconfirm" - resend confirmation code to the user for given user's identifier.

= Version 0.2.0 released 2015-05-15

Bugfix:
  * Use standard constructor of class 'ifstream'.
  * Change parameter 'p' long name from 'arg' to 'pwd'.
  * Add required include statement <cstdlib>.

Improvements:
  * Fix misspelling in the help text.
  * Add "happy path" test.
  * Modify 'encrypt' and 'verify' commands to work with Virgil PKI service thru user's identifier.

Changes:
  * Change parameter 'asn1' to 'der'.
  * Change parameters format for 'encrypt' command.

Features:
  * Add 'verify' command.
  * Add 'adduser' command

= Version 0.1.0 released 2015-05-04

Features:
  * Add command "keygen" - generate private key with given parameters.
  * Add command "key2pub" - extract public key from the private key.
  * Add command "pub2cert" - create certificate from the public key and identifiers.
  * Add command "certinfo" - output detailed information about given certificate.
  * Add command "encrypt" - encrypt data for given recipients which can be defined by certificates and by passwords.
  * Add command "decrypt" - decrypt data for given recipient which can be defined by certificate or by password.
  * Add command "sign" - sign data with private key.
  * Add command "verify" - verify data with certificate.


src/commands/virgilencrypt.cxx
/**
 * Copyright (C) 2015 Virgil Security Inc.
 *
 * Lead Maintainer: Virgil Security Inc. <support@virgilsecurity.com>
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     (1) Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *     (2) Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in
 *     the documentation and/or other materials provided with the
 *     distribution.
 *
 *     (3) Neither the name of the copyright holder nor the names of its
 *     contributors may be used to endorse or promote products derived from
 *     this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ''AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
 * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <algorithm>
#include <fstream>
#include <iostream>
#include <iterator>
#include <stdexcept>
#include <string>

#include <tclap/CmdLine.h>

#include <virgil/crypto/VirgilByteArray.h>
#include <virgil/crypto/VirgilStreamCipher.h>
#include <virgil/crypto/stream/VirgilStreamDataSource.h>
#include <virgil/crypto/stream/VirgilStreamDataSink.h>

#include <virgil/sdk/keys/model/PublicKey.h>
#include <virgil/sdk/keys/model/UserDataType.h>
#include <virgil/sdk/keys/http/Connection.h>
#include <virgil/sdk/keys/client/KeysClient.h>

#include <cli/version.h>
#include <cli/pair.h>
#include <cli/util.h>

using virgil::crypto::VirgilByteArray;
using virgil::crypto::VirgilStreamCipher;
using virgil::crypto::stream::VirgilStreamDataSource;
using virgil::crypto::stream::VirgilStreamDataSink;

using virgil::sdk::keys::model::PublicKey;
using virgil::sdk::keys::model::UserDataType;
using virgil::sdk::keys::http::Connection;
using virgil::sdk::keys::client::KeysClient;


/**
 * @brief Add recipients from the configuration files to the cipher.
 * @param configFiles - array of configuration files names.
 * @param cipher - recipients added to.
 * @return Number of added recipients.
 */
static size_t add_recipients_configs(const std::vector<std::string>& configFiles, VirgilStreamCipher* cipher);

/**
 * @brief Add recipients from the list to the cipher.
 * @param recipients - array of recipients <type:value>, where type can be [pass|vpk_file|email|phone|domain].
 * @param cipher - recipients added to.
 * @return Number of added recipients.
 */
static size_t add_recipients(const std::vector<std::string>& recipientsData, VirgilStreamCipher* cipher);

/**
 * @brief Add recipient to the cipher.
 * @param recipientData - <type:value>, where type can be [pass|key|email|phone|domain].
 * @param cipher - recipients added to.
 */
static void add_recipient(const std::string recipientIdType, const std::string recipientId,
                          VirgilStreamCipher* cipher);


#ifdef SPLIT_CLI
#define MAIN main
#else
#define MAIN encrypt_main
#endif

int MAIN(int argc, char **argv) {
    try {
        // Parse arguments.
        TCLAP::CmdLine cmd("Encrypt data", ' ', virgil::cli_version());

        TCLAP::ValueArg<std::string> inArg("i", "in", "Data to be encrypted. If omitted stdin is used.",
                false, "", "file");

        TCLAP::ValueArg<std::string> outArg("o", "out", "Encrypted data. If omitted stdout is used.",
                false, "", "file");

        TCLAP::ValueArg<std::string> contentInfoArg("c", "content-info",
                "Content info. If omitted - becomes a part of the encrypted data.",
                false, "", "file");

        TCLAP::MultiArg<std::string> recipientsConfigArg("r", "recipients",
                "File that contains information about recipients. Each line "
                "can be either empty line, or comment line, or recipient defined in format:\n"
                "[pass|file|email|phone|domain]:<value>\n"
                "where:\n"
                "\t* if pass, then <value> - recipient's password;\n"
                "\t* if file, then <value> - recipient's Virgil Public Key file stored locally;\n"
                "\t* if email, then <value> - recipient's email;\n"
                "\t* if phone, then <value> - recipient's phone;\n"
                "\t* if domain, then <value> - recipient's domain.\n",
                false, "file");

        TCLAP::UnlabeledMultiArg<std::string> recipientsArg("recipient",
                "Contains information about one recipient. "
                "Same as significant line in the recipients configuration file.",
                false, "recipient", true);

        cmd.add(recipientsArg);
        cmd.add(recipientsConfigArg);
        cmd.add(contentInfoArg);
        cmd.add(outArg);
        cmd.add(inArg);

        cmd.parse(argc, argv);

        // Create cipher
        VirgilStreamCipher cipher;

        // Add recipients
        size_t addedRecipientsCount = 0;
        addedRecipientsCount += add_recipients_configs(recipientsConfigArg.getValue(), &cipher);
        addedRecipientsCount += add_recipients(recipientsArg.getValue(), &cipher);
        if (addedRecipientsCount == 0) {
            throw std::invalid_argument("no recipients are defined");
        }

        // Prepare input
        std::istream* inStream;
        std::ifstream inFile;
        if (inArg.getValue().empty() || inArg.getValue() == "-") {
            inStream = &std::cin;
        } else {
            inFile.open(inArg.getValue(), std::ios::in | std::ios::binary);
            if (!inFile) {
                throw std::invalid_argument("can not read file: " + inArg.getValue());
            }
            inStream = &inFile;
        }

        // Prepare output
        std::ostream* outStream;
        std::ofstream outFile;
        if (outArg.getValue().empty() || outArg.getValue() == "-") {
            outStream = &std::cout;
        } else {
            outFile.open(outArg.getValue(), std::ios::out | std::ios::binary);
            if (!outFile) {
                throw std::invalid_argument("can not write file: " + outArg.getValue());
            }
            outStream = &outFile;
        }

        VirgilStreamDataSource dataSource(*inStream);
        VirgilStreamDataSink dataSink(*outStream);

        // Define whether embed content info or not
        bool embedContentInfo = contentInfoArg.getValue().empty();
        cipher.encrypt(dataSource, dataSink, embedContentInfo);

        // Write content info to file if it was not embedded
        if (!embedContentInfo) {
            std::ofstream contentInfoFile(contentInfoArg.getValue(), std::ios::out | std::ios::binary);
            if (!contentInfoFile) {
                throw std::invalid_argument("can not write file: " + contentInfoArg.getValue());
            }
            VirgilByteArray contentInfo = cipher.getContentInfo();
            std::copy(contentInfo.begin(), contentInfo.end(), std::ostreambuf_iterator<char>(contentInfoFile));
        }

    } catch (TCLAP::ArgException& exception) {
        std::cerr << "encrypt. Error: " << exception.error() << " for arg " << exception.argId() << std::endl;
        return EXIT_FAILURE;
    } catch (std::exception& exception) {
        std::cerr << "encrypt. Error: " << exception.what() << std::endl;
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}

static size_t add_recipients_configs(const std::vector<std::string>& configFiles, VirgilStreamCipher* cipher) {
    size_t addedRecipientsCount = 0;
    for (const auto& configFile : configFiles) {
        std::ifstream file(configFile);
        if (!file) {
             throw std::invalid_argument("recipientsConfigArg: can not read recipient config file: " + configFile);
        }

        // Else
        std::string recipientData;
        unsigned long long numberLine = 0;
        while (file >> std::ws && std::getline(file, recipientData)) {
            ++numberLine;
            if (!recipientData.empty() && recipientData[0] != '#') {
                const auto recipientPair = virgil::cli::parse_pair(recipientData);
                const std::string recipientIdType = recipientPair.first;
                const std::string recipientId = recipientPair.second;
                try {
                    add_recipient(recipientIdType, recipientId, cipher);
                } catch (std::exception& exception) {
                    throw std::runtime_error("can not add recipient " + recipientIdType + ":" + recipientId +
                            " from line " + std::to_string(numberLine) + " . Details: " + exception.what());
                }
               ++addedRecipientsCount;
            }
        }
    }

    return addedRecipientsCount;
}

static size_t add_recipients(const std::vector<std::string>& recipientsData, VirgilStreamCipher* cipher) {
    size_t addedRecipientsCount = 0;
    for (const auto& recipientData : recipientsData) {
        const auto recipientPair = virgil::cli::parse_pair(recipientData);
        const std::string recipientIdType = recipientPair.first;
        const std::string recipientId = recipientPair.second;
        try {
            add_recipient(recipientIdType, recipientId, cipher);
        } catch (std::exception& exception) {
            throw std::invalid_argument("can not add recipient. Error " + recipientIdType +
                    ":" + recipientId + "\n" + exception.what());
        }
        ++addedRecipientsCount;
    }
    return addedRecipientsCount;
}

static void add_recipient(const std::string recipientIdType, const std::string recipientId,
                          VirgilStreamCipher* cipher) {
    if (recipientIdType == "pass") {
        VirgilByteArray pwd = virgil::crypto::str2bytes(recipientId);
        cipher->addPasswordRecipient(pwd);
    } else {
        PublicKey publicKey;
        if (recipientIdType == "file") {
            // Read Virgil Public Key from file
            std::ifstream virgilPublicKeyFile(recipientId, std::ios::in | std::ios::binary);
            if (!virgilPublicKeyFile) {
                throw std::invalid_argument("can not read recipient's Virgil Public Key file: " + recipientId);
            }
            publicKey = virgil::cli::read_virgil_public_key(virgilPublicKeyFile);
        } else {
            // Get Virgil Public Key from the service
            publicKey = virgil::cli::get_virgil_public_key(recipientId, recipientIdType);
        }
        VirgilByteArray publicKeyId = virgil::crypto::str2bytes(publicKey.publicKeyId());
        cipher->addKeyRecipient(publicKeyId, publicKey.key());
    }
}
